osu.Server.Spectator/Hubs/Multiplayer/MultiplayerHub.cs

```diff
@@ -507,6 +507,27 @@ public async Task StartMatch()
             }
         }
 
+        public async Task AbortMatch()
+        {
+            using (var userUsage = await GetOrCreateLocalUserState())
+            using (var roomUsage = await getLocalUserRoom(userUsage.Item))
+            {
+                var room = roomUsage.Item;
+                if (room == null)
+                    throw new InvalidOperationException("Attempted to operate on a null room");
+
+                ensureIsHost(room);
+
+                foreach (var user in room.Users.Where(u => isGameplayState(u.State)))
+                {
+                    await HubContext.ChangeAndBroadcastUserState(room, user, MultiplayerUserState.Idle);
+                    await Clients.User(user.UserID.ToString()).GameplayAborted(GameplayAbortReason.HostAbortedTheMatch);
```

Comment:
I'm not sure how much it matters but rather than messaging each individual client you could also do `Clients.Group(GetGroupId(room.RoomId)).GameplayAborted(...)`?

Reply:
I think sending to all users is probably better (so we can show it as an event in the room lobby for those that are refereeing or whatever).